### PR TITLE
access_log: Added new command operator %VIRTUAL_CLUSTER_NAME% to retrieve the matched VC name

### PR DESCRIPTION
**Pull request by agrawroh**
This PR adds a new command operator in the access logs called `%VIRTUAL_CLUSTER_NAME%` which can be used to retrieve the matched Virtual Cluster name information. 

It could be very useful for understanding which Virtual Cluster is getting matched for a given route especially when we have thousands of routes and a subset of Virtual Clusters.

**Commit Message:** Added new command operator %VIRTUAL_CLUSTER_NAME% to retrieve the matched VC name
**Additional Description:** Adds a new command operator called `%VIRTUAL_CLUSTER_NAME%` in the access logs which can be used to retrieve the matched Virtual Cluster name for the incoming request. It gets populated from the Router.
**Risk Level:** Low
**Testing:** Unit Tests
**Docs Changes:** Added description on `%VIRTUAL_CLUSTER_NAME%` in the Docs.
**Release Notes:** Added
**Platform Specific Features:** N/A

**Signed-off-by:** Rohit Agrawal <rohit.agrawal@databricks.com>

### CODE REVIEW

File: source/common/formatter/substitution_formatter.cc
Comment (lambdai) [inline source/common/formatter/substitution_formatter.cc:959]: const string& ?

File: docs/root/configuration/observability/access_log/usage.rst
Comment (lambdai) [inline docs/root/configuration/observability/access_log/usage.rst:389]: This is not applied to tcp traffic. I remember there is a section define some field name are only reasonable in HTTP but not TCP but I cannot recall

**Review by lambdai — COMMENTED** [review]
Generally LGTM modulo minor nits

File: source/common/router/router.cc
Comment (lambdai) [inline source/common/router/router.cc:457]: `const std::string &`

I am uncertain if calling toSTring(statName()) is a recommended pattern in worker thread. CC @jmarantz for suggestion

File: docs/root/configuration/observability/access_log/usage.rst
Comment (agrawroh) [reply docs/root/configuration/observability/access_log/usage.rst:389]: Changed. Thanks!

File: source/common/router/router.cc
Comment (agrawroh) [reply source/common/router/router.cc:457]: @jmarantz @ggreenway Any suggestions here?

Comment (jmarantz) [reply source/common/router/router.cc]: That will take a symbol table lock. Probably not a good idea.

Does request_vcluster_ contain a string name also? It might, or it could.

Comment (agrawroh) [reply source/common/router/router.cc]: @jmarantz Thanks for the suggestion. I added a string field to store the VC name.

File: docs/root/configuration/observability/access_log/usage.rst
Comment (jmarantz) [inline docs/root/configuration/observability/access_log/usage.rst:391]: also UDP would be excluded I assume?

File: envoy/router/router.h
Comment (jmarantz) [inline envoy/router/router.h:507]: just call this `name()`?

File: envoy/stream_info/stream_info.h
Comment (jmarantz) [inline envoy/stream_info/stream_info.h:384]: rename this `virtualClusterName()` and describe what happens if no one calls setVirtualClusterName() first. Looking at population of this below. I think we should make this return `absl::optional<absl::string_view>` or `OptRef<std:string>` (see envoy/common/optref.h).

I see there's a pre-existing naming inconsistency within this class with getRouteName() above and. bytesReceived() below, but typically in our style guide we just name methods for what we are trying to get, without the explicit `get` verb.

File: source/common/formatter/substitution_formatter.cc
Comment (jmarantz) [inline source/common/formatter/substitution_formatter.cc:956]: wow this is a huge sequence of string comparisons. I'm assuming this is not in the request path. But even so maybe we could have a string->enum map and a switch statement?

not an issue for this PR but maybe leave a TODO/issue to address in a follow-up?

File: source/common/router/config_impl.h
Comment (jmarantz) [inline source/common/router/config_impl.h:263]: I think rather than doing this we should have the name be optional.

File: source/common/router/router.cc
Comment (jmarantz) [inline source/common/router/router.cc:458]: can we depend on the requst_vcluster_ outlive the stream-info? If so let's have this be a reference.

Note that the StatName() (which you previously were converting to a string) is already essentially a reference to storage held elsewhere, like a string_view or `string&`, so if we parallel the ownership of that we should be able to hold a reference to a string with a similar lifetime.

File: source/common/stream_info/stream_info_impl.h
Comment (jmarantz) [inline source/common/stream_info/stream_info_impl.h:341]: per above let's try to avoid deep-copy here.

File: envoy/router/router.h
Comment (agrawroh) [reply envoy/router/router.h:507]: Changed. Thanks!

File: docs/root/configuration/observability/access_log/usage.rst
Comment (agrawroh) [reply docs/root/configuration/observability/access_log/usage.rst:391]: Added UDP as well.

File: envoy/stream_info/stream_info.h
Comment (agrawroh) [reply envoy/stream_info/stream_info.h:384]: Changed to `virtualClusterName()` as suggested.

File: source/common/formatter/substitution_formatter.cc
Comment (agrawroh) [reply source/common/formatter/substitution_formatter.cc:956]: Added a TODO at the top as a follow-up. I'll take it up next.

File: source/common/router/router.cc
Comment (agrawroh) [reply source/common/router/router.cc:458]: @jmarantz I'm not 100% sure (not very familiar) that it'd be safe to store a reference instead of creating a Deep Copy. But if you think it's okay then I can change it.

File: source/common/router/config_impl.h
Comment (agrawroh) [reply source/common/router/config_impl.h:263]: Changed it to an optional instead. Thanks!

File: source/common/stream_info/stream_info_impl.h
Comment (agrawroh) [reply source/common/stream_info/stream_info_impl.h]: Done.

File: source/common/router/config_impl.h
Comment (jmarantz) [inline source/common/router/config_impl.h:250]: rename to `name_` and add comment that `name_ `and `stat_name_` are two different representations for the same string, retained in memory to avoid symbol-table locks that would be needed when converting on-the-fly.

File: source/common/router/router.cc
Comment (jmarantz) [inline source/common/router/router.cc:456]: `!= nullptr`

File: source/common/stream_info/stream_info_impl.h
Comment (jmarantz) [inline source/common/stream_info/stream_info_impl.h:341]: This doesn't compile, right? By default there is no virtual_cluster_name and it gets set after construction?

Try making this `OptRef<const std::string>` and having accessors to this check virtual_cluster_name_.has_value().

Also -- it's worth double-checking that we are confident that the owner for this storage outlives the StreamInfo. If not you'll need to copy the string rather than referencing it, in which case you could just use `absl::optional<std::string>`.

File: source/common/router/router.cc
Comment (jmarantz) [reply source/common/router/router.cc]: I don't have a good feel for it. If you don't either you could just leave a comment for now to investigate and do a deep copy.

**Review by jmarantz — COMMENTED** [review]
/wait

File: source/common/router/router.cc
Comment (agrawroh) [reply source/common/router/router.cc]: I changed it to do a deep copy instead for now and left a comment/TODO to investigate more for whether it could be a reference.

File: source/common/stream_info/stream_info_impl.h
Comment (agrawroh) [reply source/common/stream_info/stream_info_impl.h]: Changed it to `absl::optional<std::string>` and left a comment/TODO to investigate more for whether it could be a reference. Thanks for the suggestion.

File: docs/root/configuration/observability/access_log/usage.rst
Comment (jmarantz) [inline docs/root/configuration/observability/access_log/usage.rst:388]: nit: HTTP*/gRPC here

File: source/common/router/config_impl.h
Comment (jmarantz) [inline source/common/router/config_impl.h:239]: `const absl::optional<std::string>&`

File: source/common/stream_info/stream_info_impl.h
Comment (jmarantz) [inline source/common/stream_info/stream_info_impl.h:455]: nit -- extra space before 'string' at beginning of line.

Rather than an unowned `TODO: ` -- use `TODO(agrawroh)` or generate an issue for it and use `TODO(#ISSUE_NUMBER)`

**Review by jmarantz — COMMENTED** [review]
looks great -- just a few minor nits

File: docs/root/configuration/observability/access_log/usage.rst
Comment (agrawroh) [reply docs/root/configuration/observability/access_log/usage.rst:388]: Done, Thanks!

File: source/common/stream_info/stream_info_impl.h
Comment (agrawroh) [reply source/common/stream_info/stream_info_impl.h:455]: Changed it to `TODO(agrawroh)`.

**Review by jmarantz — DISMISSED** [review]
/assign-from @envoyproxy/senior-maintainers

File: source/common/formatter/substitution_formatter.cc
Comment (zuercher) [inline source/common/formatter/substitution_formatter.cc:964]: I believe FieldExtractor's return value is an optional string and we should return `absl::null_opt` here. In fact, I think the entire function body can just be `return stream_info.virtualClusterName()`.

This allows us to distinguish whether the value is set or not (e.g. `absl::null_opt`). The decision whether to render a hyphen for unset values is left to code that understands the context and whether we need a `ProtobufWkt::Value` (which can encode a null value) or a string that defaults to `"-"`.

**Review by zuercher — CHANGES_REQUESTED** [review]
Sorry for the slow response, I was out last week. Just one comment below.

File: source/common/formatter/substitution_formatter.cc
Comment (agrawroh) [reply source/common/formatter/substitution_formatter.cc]: @zuercher Thanks for taking a look. I changed it to return an optional as you suggested.

**Review by zuercher — APPROVED** [review]
Thanks!

**Pull request merged**
